docs/vetomint.md

```diff
@@ -0,0 +1,196 @@
+proposal or against it), it immediately stops and casts a precommit based on the
+votes it has collected so far. In the following paragraphs, we explain the
+rationale behind this modification. We will also illustrate that the choice of
+5/6 threshold naturally implies the byzantine threshold of 1/6.
+
+Let's recall the relevant behavior of Tendermint to understand this decision. In
+Tendermint, a validator in the prevote phase can move to the precommit phase,
+skipping a timeout event, if it collects more than 2/3 of the accepting prevotes
+(i.e., prevotes to the proposal, or non-nil prevotes) in the given round.
+Assuming that all non-byzantine validators participate promptly without network
+delay, this significantly optimizes the consensus. However, Tendermint does not
+allow early termination based on the total number of prevotes received, even if
+more than 2/3 of the prevotes (including nil prevotes) have been collected. This
+behavior is not problematic in the original Tendermint because the timeout
+interval is typically short and votes from non-byzantine validators do not
+split, as the acceptance of proposals are decided deterministically by a set of
+predefined rules.
+
+Since we have assumed that validators in Vetomint may freely cast nil prevotes,
+however, the lack of *early termination* becomes a severe
+issue. For example, it might be the case where the half of validators accept a
+proposal while the other half reject it. In such a case, in the original
+Tendermint, although all validators have make decisions, validators can do
+nothihg but wait for the pending timeout expiration, which can be intolerably
+long.
+
+To prevent such issues, we add a new early termination rule to Tendermint. If a
+validator collects more than 5/6 of the total prevotes, either accepting ones or
+rejecting ones, it makes progress to the next phase, depending on the prevotes
+it has received so far. That is, among the received prevotes, if accepting
+prevotes are more than 2/3 of the total prevotes, the validator broadcasts a
+precommit to the proposal (i.e., an accepting precommit), and broadcasts a nill
+precommit otherwise.
+
+The threshold of 5/6 is chosen carefully. To demonstrate this, consider the
+following extreme case where the threshold is 2/3 (thresholds less than 2/3 or
+greater than 1 are meaningless). Suppose that an inarguably acceptable block is
+proposed, and all but one malicious validator, who is trying to delay the
+consensus, agree on the proposal. In the desirable scenario, the malicious
+action should be effectively ignored. However, we can easly see that the
+malicious validator has a high probability of hindering the consensus simply by
+broadcasting a nil prevote. All validators that receive the nil prevote before
+the early termination will (conservatively) assume that a quorum has been failed
+to met (because the received accepting prevotes are less than 2/3). In short,
```

Comment:
I quite can't fully understand the reasoning in this sentence. I thought early termination only considers the total prevotes, not whether it is accepting or not.

Reply:
Discussed in Discord, well understood.
https://discord.com/channels/941206496203669545/1049619693020905542/1064427038687121479